arbos/l2pricing/model.go

```diff
@@ -18,20 +18,48 @@ const InitialMinimumBaseFeeWei = params.GWei / 10
 const InitialBaseFeeWei = InitialMinimumBaseFeeWei
 const InitialGasPoolSeconds = 10 * 60
 const InitialRateEstimateInertia = 60
+const InitialExponentialMechanismDenom = 102
+const InitialExponentialMechanismTolerance = 10
 
 var InitialGasPoolTargetBips = arbmath.PercentToBips(80)
 var InitialGasPoolWeightBips = arbmath.PercentToBips(60)
 
-func (ps *L2PricingState) AddToGasPool(gas int64) error {
+const FirstExponentialPricingVersion = 4
+
+func (ps *L2PricingState) AddToGasPool(gas int64, arbosVersion uint64) error {
 	gasPool, err := ps.GasPool()
 	if err != nil {
 		return err
 	}
+	if arbosVersion >= FirstExponentialPricingVersion && gasPool >= 0 {
+		gasPool = 0
+	}
 	return ps.SetGasPool(arbmath.SaturatingAdd(gasPool, gas))
 }
 
 // Update the pricing model with info from the last block
 func (ps *L2PricingState) UpdatePricingModel(l2BaseFee *big.Int, timePassed uint64, arbosVersion uint64, debug bool) {
+	if arbosVersion < FirstExponentialPricingVersion {
+		// note: if we restart the chain at version >= FirstExponentialPricingVersion, we can simplify the L2-pricing-related params and precompiles
+		ps.UpdatePricingModelPreExpVersion(l2BaseFee, timePassed, arbosVersion, debug)
+		return
+	}
+
+	speedLimit, _ := ps.SpeedLimitPerSecond()
+	_ = ps.AddToGasPool(int64(timePassed*speedLimit), arbosVersion)
+	denom, _ := ps.ExponentialMechanismDenom()
+	tolerance, _ := ps.ExponentialMechanismTolerance()
+	gasPool, _ := ps.GasPool()
+	minBaseFee, _ := ps.MinBaseFeeWei()
+	baseFee := minBaseFee
+	if gasPool+int64(tolerance*speedLimit) < 0 {
```

Comment:
If the gas pool is intended to be non-positive, we'll need to figure out what to do with `PerBlockGasLimit()`, which right now always return `0` and leave all L2 blocks unfilled

Reply:
Fixed this so that `PerBlockGasLimit` returns the default value when we're in exponential mechanism.